packages/rx_bloc_cli/mason_templates/bricks/lib_auth_matrix/__brick__/lib/lib_auth_matrix/services/auth_matrix_service.dart

```diff
@@ -0,0 +1,173 @@
+{{> licence.dart }}
+
+import 'package:rxdart/rxdart.dart';
+import 'package:uuid/uuid.dart';
+import 'package:widget_toolkit_otp/widget_toolkit_otp.dart';
+import 'package:widget_toolkit_pin/widget_toolkit_pin.dart';
+
+import '../../base/models/errors/error_model.dart';
+import '../../lib_router/router.dart';
+import '../../lib_router/services/router_service.dart';
+import '../models/action_request.dart';
+import '../models/auth_matrix_action_type.dart';
+import '../models/auth_matrix_cancel_model.dart';
+import '../models/auth_matrix_response.dart';
+import '../models/auth_matrix_verify.dart';
+import '../repositories/auth_matrix_repository.dart';
+
+/// Service that handles the auth matrix flow,
+class AuthMatrixService implements SmsCodeService, PinCodeService {
+  AuthMatrixService(
+    this._authMatrixRepository,
+    this._routerService,
+  );
+
+  final RouterService _routerService;
+  final AuthMatrixRepository _authMatrixRepository;
+  final BehaviorSubject<AuthMatrixResponse> _flowEvents =
+      BehaviorSubject<AuthMatrixResponse>();
+
+  String _endToEndId = '';
+  String _userData = '';
+
+  ///Function used to generate the end to end id
+  String generateEndToEndId() => const Uuid().v4();
+
+  ///Function used to initialise the auth matrix flow, 
+  ///and navigate to the correct screen based on the server response
+  Future<void> initialiseAuthMatrixFlow({
+    required ActionRequest payload,
+    required String userData,
+  }) async {
+    _flowEvents
+        .add(await _authMatrixRepository.initiateAuthMatrix(payload: payload));
+    _endToEndId = payload.endToEndId;
+    _userData = userData;
+    await for (AuthMatrixResponse response in _flowEvents) {
+      switch (response.authZ) {
+        case AuthMatrixActionType.none:
+          _routerService.pop();
+          return;
+        case AuthMatrixActionType.pinAndOtp:
+        case AuthMatrixActionType.pinOnly:
+          await _navigateFunction(response, payload.endToEndId);
+      }
+    }
+    return;
+  }
+
+  ///Function used to navigate the user to the correct screen
+  /// based on the server response
+  Future<void> _navigateFunction(
+      AuthMatrixResponse response, String endToEndId) async {
+    switch (response.authZ) {
+      case AuthMatrixActionType.pinOnly:
+        await _routerService.go(
+            AuthMatrixPinBiometricsRoute(
+              endToEndId,
+            ),
+            response);
+      case AuthMatrixActionType.pinAndOtp:
+        await _routerService.go(
+            AuthMatrixOtpRoute(
+              endToEndId,
+            ),
+            response);
+      case AuthMatrixActionType.none:
+        break;
+    }
```

Comment:
We can rewrite this so that we use the `_navigateFunction()` inside the `await for` loop for all the switch cases.

Currently we are only using the await for loop for making a distinction between no auth matrix action (`none`) and the rest of them. Any other action is handled in the _navigateFunction, while the none action is ignored there. Since all actions use `_routerService` for some kind of behaviour, we can unify them in one place.

Reply:
Done